cmake/googletest.cmake

```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Fangjun Kuang (csukuangfj@gmail.com)
+# See ../COPYING for clarification regarding multiple authors
+
+function(download_googltest)
+  include(ExternalProject)
+
+  set(googletest_URL  "https://github.com/google/googletest/archive/release-1.10.0.tar.gz")
```

Comment:
so we prefer downloading 3rd party libraries to adding them as submodules? Just a question.

Reply:
As far as I am concerned, `ExternalProject` makes developers and users' life simpler
than `git submodule`.

We do not need to know how to use `git submodule` to add, init, update, or remove
submodules. All we need to do is to run 

```bash
cd build
cmake ..
make
```

`cmake` will download all dependencies for us.

`MACE` from xiaomi also uses `ExternalProject`. I think you can also ask your colleagues
why they prefer `ExternalProject` over `git submodule`.

Reply:
ok, thanks

Reply:
On the contrary, ExternalProject is not recommended by Craig Scott in his book "Professional CMake". Basically, the problem with ExternalProject is that it downloads and configures the third-party library at build time, not configure time. This means that your top-level project cannot access the variables and targets declared by the third-party library.

Instead, you are forced to define the variables of the built targets on your own, which is unfortunately fragile. I know it will never change in the case of gtest, but these things can change more often for active projects.

You can see alternatives here: https://cliutils.gitlab.io/modern-cmake/chapters/testing/googletest.html

I basically always use git submodules for my own work. FetchContent is basically a better ExternalProject, if you don't want submodules. Take a look at that page.

By the way, you should be able to target the latest cmake instead of cmake 3.5 freely. It's very easy to install cmake nowadays as a binary. You can literally do `pip install cmake` and get the latest cmake now! (The Modern CMake webpage discusses this.) FetchContent's CMake 3.11 dependency is therefore no problem.

Reply:
@galv
thanks for the info.

Maybe we can combine `ExternalProject_Add` and `git submodule` as what PyTorch
is currently doing:

https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L413
> include(ExternalProject)

https://github.com/pytorch/pytorch/blob/master/.gitmodules

----

As for the CMake version, I would like to use the latest CMake in my own projects, since it provides
more functionally. For example,  `FetchContent` is only available since CMake 3.11.

I know it is very easy for me and you to install CMake and it does not require `sudo` permission to install it.  Users of an open-source project may have various background
and some of them may have difficulty or be reluctant to install it.

I guess not using the latest CMake may be more portable for an open-source project.

I find that PyTorch also uses CMake 3.5 (https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L1)
> cmake_minimum_required(VERSION 3.5 FATAL_ERROR)

Facebooks' Folly even uses CMake 3.0 (https://github.com/facebook/folly/blob/97abe3c96a4061322b6bb4da8042540657e1a3f7/CMakeLists.txt#L15)
> cmake_minimum_required(VERSION 3.0.2 FATAL_ERROR)

-----

PS: you can use `FetchContent` in CMake 3.5 if you like. What you need to do
is just copy `Modules/FetchContent.cmake` and `ModulesFetchContent/CMakeLists.cmake.in` from CMake 3.11 and place them in your projects.